src/api.rs

```diff
@@ -9,145 +14,235 @@ use hyper::{Response, StatusCode};
+        .and(with_store.clone())
+        .and_then(
+            move |id: String,
+                  idempotency_key: Option<String>,
+                  body: warp::body::FullBody,
+                  engine: E,
+                  store: S| {
+                // Gets called by our settlement engine, forwards the request outwards
+                // until it reaches the peer's settlement engine.
+                let message = Vec::from_buf(body);
+                let input = format!("{}{:?}", id, message);
+                let input_hash = get_hash_of(input.as_ref());
+
+                // Wrap do_send_outgoing_message in a closure to be invoked by
+                // the idempotency wrapper
+                let receive_message_fn = move || engine.receive_message(id, message);
+                make_idempotent_call(store, receive_message_fn, input_hash, idempotency_key)
+                    // TODO Replace with error case with response
+                    .map_err(move |(_status_code, error_msg)| warp::reject::custom(error_msg))
+                    .and_then(move |(status_code, message)| {
+                        Ok(Response::builder()
+                            .status(status_code)
+                            .body(message)
+                            .unwrap())
+                    })
+            },
+        );
 
-        #[post("/accounts/:account_id/messages")]
-        /// Forwards the data to the API engine's `receive_message` function.
-        /// Endpoint: POST /accounts/:id/messages
-        fn receive_message(&self, account_id: String, body: Vec<u8>, idempotency_key: Option<String>) -> impl Future<Item = Response<String>, Error = Response<String>> {
-            let input = format!("{}{:?}", account_id, body);
-            let input_hash = get_hash_of(input.as_ref());
-            let account_id = account_id.clone();
-            let engine = self.engine.clone();
-            let f = move || engine.receive_message(account_id, body);
-            self.make_idempotent_call(f, input_hash, idempotency_key)
-        }
+    accounts.or(settlements).or(messages).boxed()
+}
 
-        #[post("/accounts")]
-        /// Forwards the data to the API engine's `create_account` function.
-        /// Endpoint: POST /accounts/
-        fn create_account(&self, body: CreateAccount, idempotency_key: Option<String>) -> impl Future<Item = Response<String>, Error = Response<String>> {
-            let input = format!("{:?}", body);
-            let input_hash = get_hash_of(input.as_ref());
-            let engine = self.engine.clone();
-            let f = move || engine.create_account(body);
-            self.make_idempotent_call(f, input_hash, idempotency_key)
-        }
+// Helper function that returns any idempotent data that corresponds to a
+// provided idempotency key. It fails if the hash of the input that
+// generated the idempotent data does not match the hash of the provided input.
+fn check_idempotency<S>(
```

Comment:
If we're going to make this crate depend on `interledger-api`, what do you think about having that crate export shared functionality like this (also behind a feature flag)?

Reply:
I agree with that.